app/src/main/java/com/techcourse/service/UserService.java

```diff
@@ -1,32 +1,65 @@
 package com.techcourse.service;
 
+import com.interface21.dao.DataAccessException;
+import com.techcourse.config.DataSourceConfig;
 import com.techcourse.dao.UserDao;
 import com.techcourse.dao.UserHistoryDao;
 import com.techcourse.domain.User;
 import com.techcourse.domain.UserHistory;
+import java.sql.Connection;
+import java.sql.SQLException;
+import javax.sql.DataSource;
 
 public class UserService {
 
     private final UserDao userDao;
     private final UserHistoryDao userHistoryDao;
+    private final DataSource dataSource;
 
     public UserService(final UserDao userDao, final UserHistoryDao userHistoryDao) {
         this.userDao = userDao;
         this.userHistoryDao = userHistoryDao;
+        this.dataSource = DataSourceConfig.getInstance();
     }
 
-    public User findById(final long id) {
-        return userDao.findById(id);
+    public User findById(final Connection conn, final long id) {
+        return userDao.findById(conn, id);
     }
 
-    public void insert(final User user) {
-        userDao.insert(user);
-    }
+    // todo : Connection 리팩토링 이후 주석 삭제 예정
+//    public void insert(final User user) {
+//        userDao.insert(user);
+//    }
 
     public void changePassword(final long id, final String newPassword, final String createBy) {
-        final var user = findById(id);
-        user.changePassword(newPassword);
-        userDao.update(user);
-        userHistoryDao.log(new UserHistory(user, createBy));
+        Connection conn = null;
+        try {
+            conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
+
+            final var user = findById(conn, id);
+            user.changePassword(newPassword);
+            userDao.update(conn, user);
+
+            userHistoryDao.log(conn, new UserHistory(user, createBy));
+
+            conn.commit();
+        } catch (DataAccessException | SQLException e) {
+            if (conn != null) {
+                try {
+                    conn.rollback();
+                } catch (SQLException rollbackEx) {
+                    throw new DataAccessException("비밀번호 변경 롤백 실패", rollbackEx);
+                }
+            }
+            throw new DataAccessException("비밀번호 변경 실패", e);
+        } finally {
+            if (conn != null) {
+                try {
+                    conn.close();
+                } catch (SQLException ignored) {
+                }
+            }
+        }
```

Comment:
```suggestion
        try (Connection conn = dataSource.getConnection()) {
            conn.setAutoCommit(false);
            try {
                final var user = findById(conn, id);
                user.changePassword(newPassword);
                userDao.update(conn, user);
                userHistoryDao.log(conn, new UserHistory(user, createBy));

                conn.commit();
            } catch (Exception e) {
                conn.rollback();
                throw e;
            }
        } catch (DataAccessException | SQLException e) {
            throw new DataAccessException("비밀번호 변경 실패", e);
        }
```

이렇게 try catch 범위를 좁혀서 내부를 transaction 및 rollback에 집중하도록 하는 거 어떤가요!?
Connection을 외부에서 null로 선언 및 초기화 안해줘도 되고, try wtih resource 사용으로 가독성 향상과 안정성이 높아질 수 있을 것 같습니다.
물론 제 코드가 정답은 아니니 참고만 해주세용!!

그리고 현재 코드는 만약 (지금은 그럴일은 없겠으나) DataAccessException, SQLException 이 아닌 다른 Exception이 터졌을 때 rollback이 되지 않을 경우도 있을 것 같아요.

Reply:
이중으로 처리해 리팩토링하는 거 매우 좋았습니다! 👍👍👍👍👍👍👍👍👍👍👍👍

반영했습니다! ☺️